src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs

```diff
@@ -49,28 +62,16 @@ internal sealed partial class LocalRewriter
 
             statementList.AddRange(existingStatements);
             return statementList.ToImmutableAndFree();
-
         }
 
-        private static BoundStatement ConstructIfStatementForParameter(ParameterSymbol parameter, SyntheticBoundNodeFactory factory)
+        private static BoundStatement ConstructNullCheck(ParameterSymbol parameter, MethodSymbol throwIfNullMethod, SyntheticBoundNodeFactory factory)
         {
-            BoundExpression paramIsNullCondition;
-            var loweredLeft = factory.Parameter(parameter);
-
-            if (loweredLeft.Type.IsNullableType())
-            {
-                paramIsNullCondition = factory.Not(factory.MakeNullableHasValue(loweredLeft.Syntax, loweredLeft));
-            }
-            else
-            {
-                paramIsNullCondition = factory.MakeNullCheck(loweredLeft.Syntax, loweredLeft, BinaryOperatorKind.Equal);
-            }
-
-            var argumentName = ImmutableArray.Create<BoundExpression>(factory.StringLiteral(parameter.Name));
-            BoundObjectCreationExpression ex = factory.New(factory.WellKnownMethod(WellKnownMember.System_ArgumentNullException__ctorString), argumentName);
-            BoundThrowStatement throwArgNullStatement = factory.Throw(ex);
-
-            return factory.HiddenSequencePoint(factory.If(paramIsNullCondition, throwArgNullStatement));
+            var call = factory.Call(
+                receiver: null,
+                throwIfNullMethod,
+                arg0: factory.Convert(factory.SpecialType(SpecialType.System_Object), factory.Parameter(parameter)),
```

Comment:
This will box nullable value types. That doesn't feel like an optimization.

Reply:
This is true. However we considered nullable value types a pathological scenario to the point where I judged keeping the previous code path around for it was not worthwhile. This is also why we give a warning in this scenario.

Given the similar concerns about pointers (preferable to use "existing" codegen) perhaps we may want to restore the behavior. Or generate and use only the "Throw" helper in this case and reserve the "ThrowIfNull" helper for things which are known to be reference types.

Reply:
It does seem odd. I can try and convince myself of scenarios, but why would I write a method that accepts a nullable value type, only to throw if it's null. You'd just make it take the value type and force the caller to handle it being nullable. Like, we have a compile time check for this, when would I ever want to fallback to a runtime check instead?